certificateshop-logic/src/main/java/com/epam/esm/dao/impl/jpa/JpaCertificateDaoImpl.java

```diff
@@ -0,0 +1,205 @@
+            " LIMIT ?, ?) as query1)";
+    private static final List<String> CERTIFICATE_HEADERS = Arrays.asList(ColumnNames.TABLE_GIFT_CERTIFICATE_COLUMN_ID,
+            ColumnNames.TABLE_GIFT_CERTIFICATE_COLUMN_NAME, ColumnNames.TABLE_GIFT_CERTIFICATE_COLUMN_DESCRIPTION,
+            ColumnNames.TABLE_GIFT_CERTIFICATE_COLUMN_DURATION, ColumnNames.TABLE_GIFT_CERTIFICATE_COLUMN_CREATE_DATE,
+            ColumnNames.TABLE_GIFT_CERTIFICATE_COLUMN_PRICE, ColumnNames.TABLE_GIFT_CERTIFICATE_COLUMN_LAST_UPDATE_DATE,
+            ColumnNames.TABLE_TAG_COLUMN_ID, ColumnNames.TABLE_TAG_COLUMN_NAME);
+
+    private GiftCertificateRepository giftCertificateRepository;
+    private EntityManager entityManager;
+    private ListToSetConverter listToSetConverter;
+    private GiftCertificateExtractor giftCertificateExtractor;
+
+    @Autowired
+    public JpaCertificateDaoImpl(GiftCertificateRepository giftCertificateRepository, EntityManager entityManager,
+                                 ListToSetConverter listToSetConverter, GiftCertificateExtractor giftCertificateExtractor) {
+        this.giftCertificateRepository = giftCertificateRepository;
+        this.entityManager = entityManager;
+        this.listToSetConverter = listToSetConverter;
+        this.giftCertificateExtractor = giftCertificateExtractor;
+    }
+
+    /**
+     * Saves {@link GiftCertificate} in the database.
+     *
+     * @param entity is the {@link GiftCertificate} to save.
+     */
+    @Override
+    public void save(GiftCertificate entity) {
+        entityManager.persist(entity);
+
+//        giftCertificateRepository.save(entity);
+    }
+
+    /**
+     * Finds all {@link GiftCertificate} entity in the database.
+     *
+     * @param pageNumber              is the pageNumber query parameter.
+     * @param amountEntitiesOnThePage is the amountEntitiesOnThePage query parameter.
+     * @return List of the {@link GiftCertificate} objects.
+     */
+    @Override
+    public List<GiftCertificate> findAllPagination(int pageNumber, int amountEntitiesOnThePage) {
+        Query query = entityManager.createNativeQuery(FIND_ALL_ENTITIES_SQL_PAGINATION)
+                .setParameter(1, pageNumber * amountEntitiesOnThePage)
+                .setParameter(2, amountEntitiesOnThePage);
+        List<Object[]> resultList = query.getResultList();
+        List<List<Object>> result = convertListOfArrayToListOfLists(resultList);
+        return getEntities(result);
+    }
+
+    private List<GiftCertificate> getEntities(List<List<Object>> result) {
+        ResultSet resultSet;
+        List<GiftCertificate> certificates = new ArrayList<>();
+        try {
+            resultSet = listToSetConverter.getResultSet(CERTIFICATE_HEADERS, result);
+            certificates = giftCertificateExtractor.extractData(resultSet);
+        } catch (Exception exception) {
+            throw new RuntimeException(exception);
+        }
```

Comment:
When can the mentioned exception be thrown?

Reply:
If smth. go wrong inside converter. Ii just catches this fact and throws not checked exception but unchecked.